integration/duckdb/src/lance/duckdb/lance_reader.cc

```diff
@@ -0,0 +1,307 @@
+#include <fmt/ranges.h>
+#include <lance/arrow/dataset.h>
+
+#include <cstdio>
+#include <duckdb/parser/expression/constant_expression.hpp>
+#include <duckdb/parser/expression/function_expression.hpp>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "lance/arrow/type.h"
+#include "lance/duckdb/lance.h"
+
+namespace lance::duckdb {
+
+namespace {
+
+// Forward declaration
+void ArrowArrayToVector(const std::shared_ptr<::arrow::Array> &arr, ::duckdb::Vector *out);
+
+struct GlobalScanState : public ::duckdb::GlobalTableFunctionState {
+  std::shared_ptr<lance::arrow::LanceDataset> dataset;
+  ::arrow::dataset::TaggedRecordBatchGenerator batch_generator;
+};
+
+struct LocalScanState : public ::duckdb::LocalTableFunctionState {};
+
+/// BindData for Lance Scan
+struct ScanBindData : public ::duckdb::TableFunctionData {
+  std::shared_ptr<lance::arrow::LanceDataset> dataset;
+};
+
+std::unique_ptr<::duckdb::FunctionData> LanceScanBind(
+    ::duckdb::ClientContext &context,
+    ::duckdb::TableFunctionBindInput &input,
+    std::vector<::duckdb::LogicalType> &return_types,
+    std::vector<std::string> &names) {
+  auto dataset_uri = input.inputs[0].GetValue<std::string>();
+  std::string path;
+  auto fs = GetResult(::arrow::fs::FileSystemFromUriOrPath(dataset_uri, &path));
+  auto dataset = GetResult(lance::arrow::LanceDataset::Make(std::move(fs), path));
+  auto schema = dataset->schema();
+  auto bind_data = std::make_unique<ScanBindData>();
+  bind_data->dataset = std::move(dataset);
+  for (int i = 0; i < schema->fields().size(); ++i) {
+    const auto &field = schema->field(i);
+    names.emplace_back(field->name());
+    return_types.emplace_back(ToLogicalType(*field->type()));
+    bind_data->column_ids.emplace_back(i);
+  }
+  return std::move(bind_data);
+}
+
+std::unique_ptr<::duckdb::GlobalTableFunctionState> InitGlobal(
+    ::duckdb::ClientContext &context, ::duckdb::TableFunctionInitInput &input) {
+  auto bind_data = dynamic_cast<const ScanBindData *>(input.bind_data);
+  assert(bind_data != nullptr);
+
+  auto state = std::make_unique<GlobalScanState>();
+  state->dataset = bind_data->dataset;
+
+  auto schema = state->dataset->schema();
+  std::vector<std::string> columns;
+  for (auto &column_id : input.column_ids) {
+    columns.emplace_back(schema->field(column_id)->name());
+  }
+
+  auto builder = GetResult(state->dataset->NewScan());
+  CheckStatus(builder->Project(columns));
+  auto scanner = GetResult(builder->Finish());
+  state->batch_generator = GetResult(scanner->ScanBatchesAsync());
+  return state;
+}
+
+/// Convert numeric array to duckdb vector.
+template <ArrowType T>
+void ToVector(const std::shared_ptr<::arrow::Array> &arr, ::duckdb::Vector *out) {
+  // TODO: dynamic_pointer_cast does not work here, IDK why.
+  auto array = std::static_pointer_cast<typename ::arrow::TypeTraits<T>::ArrayType>(arr);
+  assert(array != nullptr);
+  // TODO: How to use zero copy to move data from arrow to duckdb.
```

Comment:
The duckdb-pyarrow integration is zero-copy right? i forgot how but vaguely remember seeing code that uses the Arrow C structs to dress up arrow memory buffer as duckdb vector. not a blocker for this tho i think

Reply:
I checked a few places, i.e.

* `pythonpkg` https://github.com/duckdb/duckdb/blob/master/tools/pythonpkg/src/include/duckdb_python/array_wrapper.hpp
* https://github.com/duckdb/duckdb/blob/master/src/include/duckdb/common/arrow/arrow_appender.hpp

Seems no code that convert from arrow => duckdb, it only has the reversed conversion (i..e, duckdb => arrow). 

I can look into it more later. Filed https://github.com/eto-ai/lance/issues/348 to track the work.